File: docx_parser_converter/docx_to_txt/txt_generator.py
from docx_parser_converter.docx_parsers.models.paragraph_models import Paragraph
from docx_parser_converter.docx_parsers.models.document_models import DocumentSchema
from docx_parser_converter.docx_parsers.models.table_models import Table
from docx_parser_converter.docx_to_txt.converters.paragraph_converter import ParagraphConverter
from docx_parser_converter.docx_to_txt.converters.table_converter import TableConverter


class TxtGenerator:
    """
    Class to generate plain text from the document schema.
    """

    @staticmethod
    def generate_txt(document_schema: DocumentSchema, numbering_schema, indent: bool) -> str:
        """
        Generate plain text from the document schema.

        Args:
            document_schema (DocumentSchema): The document schema.
            numbering_schema: The numbering schema.
            indent (bool): Whether to apply indentation.

        Returns:
            str: Plain text representation of the document.

        Example:
            .. code-block:: python

                txt_content = TxtGenerator.generate_txt(document_schema, numbering_schema, indent=True)
        """
        body_text = TxtGenerator.generate_txt_body(document_schema.doc_margins, document_schema.elements, numbering_schema, indent)
        return body_text

    @staticmethod
    def generate_txt_body(doc_margins, elements, numbering_schema, indent: bool) -> str:
        """
        Generate the body text from document elements.

        Args:
            doc_margins: The document margins.
            elements: The document elements.
            numbering_schema: The numbering schema.
            indent (bool): Whether to apply indentation.

        Returns:
            str: Body text as a string.

        Example:
            .. code-block:: python

                body_text = TxtGenerator.generate_txt_body(doc_margins, elements, numbering_schema, indent=True)
        """
        body = ""
        if doc_margins:
            # Handle document margins if needed
            pass

        prev_paragraph = None
        for element in elements:
            if isinstance(element, Paragraph):
                if prev_paragraph:
                    body += ParagraphConverter.add_spacing(prev_paragraph, element)
                paragraph_text = ParagraphConverter.convert_paragraph(element, numbering_schema, indent)
                body += paragraph_text + "\n"
                prev_paragraph = element
            elif isinstance(element, Table):
                table_text = TableConverter.convert_table(element, numbering_schema, indent)
                body += table_text
                prev_paragraph = None  # Reset previous paragraph after a table

        return body

File: docx_parser_converter_ts/src/docx_to_txt/txtGenerator.ts
import type { DocumentSchema, DocumentElement } from '../docx_parsers/models/documentModels';
import type { NumberingSchema } from '../docx_parsers/models/numberingModels';
import type { Paragraph } from '../docx_parsers/models/paragraphModels';
import type { Table } from '../docx_parsers/models/tableModels';
import { ParagraphConverter } from './converters/paragraphConverter';
import { TableConverter } from './converters/tableConverter';
import { NumberingConverter } from './converters/numberingConverter';

/**
 * Generates plain text from the document schema, preserving the logic of the Python implementation.
 */
export class TxtGenerator {
    /**
     * Generate plain text from the document schema.
     * @param documentSchema The document schema.
     * @param numberingSchema The numbering schema.
     * @param indent Whether to apply indentation.
     * @returns Plain text representation of the document.
     */
    public static generateTxt(documentSchema: DocumentSchema, numberingSchema: NumberingSchema, indent: boolean): string {
        NumberingConverter.resetCounters();
        return TxtGenerator.generateTxtBody(documentSchema.elements, numberingSchema, indent);
    }

    /**
     * Generate the body text from document elements.
     * @param elements The document elements.
     * @param numberingSchema The numbering schema.
     * @param indent Whether to apply indentation.
     * @returns Body text as a string.
     */
    public static generateTxtBody(elements: DocumentElement[], numberingSchema: NumberingSchema, indent: boolean): string {
        let body = '';
        let prevParagraph: Paragraph | null = null;
        for (const element of elements) {
            if (TxtGenerator.isParagraph(element)) {
                if (prevParagraph) {
                    body += ParagraphConverter.addSpacing(prevParagraph, element);
                }
                const paragraphText = ParagraphConverter.convertParagraph(element, numberingSchema, indent);
                body += paragraphText + '\n';
                prevParagraph = element;
            } else if (TxtGenerator.isTable(element)) {
                const tableText = TableConverter.convertTable(element, numberingSchema, indent);
                body += tableText;
                prevParagraph = null; // Reset previous paragraph after a table
            }
        }
        return body;
    }

    private static isParagraph(element: DocumentElement): element is Paragraph {
        // TypeScript type guard for Paragraph
        return (element as Paragraph).runs !== undefined;
    }

    private static isTable(element: DocumentElement): element is Table {
        // TypeScript type guard for Table
        return (element as Table).rows !== undefined;
    }
} 